doc/src/manual/interfaces.md

```diff
@@ -371,3 +371,190 @@ If you are defining an array type that allows non-traditional indexing (indices
+the main task is the allocation of an appropriate result object.
+(This is not an issue for `broadcast!`, where
+the result object is passed as an argument.) This process is split into two stages: computation
+of the type from the arguments ([`Broadcast.rule`](@ref)), and allocation of the object
+given the resulting type with a broadcast-specific [`similar`](@ref).
+
+`Broadcast.rule` is somewhat analogous to [`promote_rule`](@ref), except that you
+may only need to define a unary variant. The unary variant simply states that you intend to
+handle broadcasting for this type, and do not wish to rely on the default fallback. Most
+implementations will be simple:
+
+```julia
+Broadcast.rule(::Type{<:MyType}) = MyType
+```
+where unary `rule` should typically discard type parameters so that any binary `rule` methods
+can be concrete (without using `<:` for type arguments).
+
+For `AbstractArray` types, this prevents the fallback choice, `Broadcast.BottomArray`,
+which is an `AbstractArray` type that "loses" to every other `AbstractArray` type in a binary call
+`Broadcast.rule(S, T)` for two types `S` and `T`.
+You do not need to write a binary `rule` unless you want to establish precedence for
+two or more non-`BottomArray` types. If you do write a binary rule, you do not need to
+supply the types in both orders, as internal machinery will try both. For more detail,
+see [below](@ref writing-binary-broadcasting-rules).
+
+The actual allocation of the result array is handled by specialized implementations of `similar`:
+
+```julia
+Base.similar(f, r::Broadcast.Result{ContainerType}, As...)
+```
+
+`f` is the operation being performed and `ContainerType` signals the resulting
+container type (e.g., `Broadcast.BottomArray`, `Tuple`, etc.).
+`eltype(r)` returns the element type, and `indices(r)` the object's indices.
+`As...` is the list of input objects. You may not need to use `f` or `As...`
+unless they help you build the appropriate object; the fallback definition is
+
+```julia
+Base.similar(f, r::Broadcast.Result{BottomArray}, As...) = similar(Array{eltype(r)}, indices(r))
+```
+
+However, if needed you can specialize on any or all of these arguments.
+
+For a complete example, let's say you have created a type, `ArrayAndChar`, that stores an
+array and a single character:
+
+```jldoctest
+struct ArrayAndChar{T,N} <: AbstractArray{T,N}
+    data::Array{T,N}
+    char::Char
+end
+Base.size(A::ArrayAndChar) = size(A.data)
+Base.getindex(A::ArrayAndChar{T,N}, inds::Vararg{Int,N}) where {T,N} = A.data[inds...]
+Base.setindex!(A::ArrayAndChar{T,N}, val, inds::Vararg{Int,N}) where {T,N} = A.data[inds...] = val
+Base.showarg(io::IO, A::ArrayAndChar, toplevel) = print(io, typeof(A), " with char '", A.char, "'")
+```
+
+You might want broadcasting to preserve the `char` "metadata." First we define
+
+```jldoctest
+Broadcast.rule(::Type{AC}) where AC<:ArrayAndChar = ArrayAndChar
+```
+
+This forces us to also define a `similar` method:
```

Comment:
Why not fall back to calling `similar(ContainerType, eltype(r), indices(r))`?

Reply:
We don't support a type for the first argument of this syntax.

Reply:
Ah, right. But I think we should, see https://github.com/JuliaLang/julia/pull/20815. This PR is another good argument to do that, since that would avoid defining two almost identical `similar` functions in simple cases.